zerver/views/registration.py

```diff
@@ -877,6 +881,56 @@ def redirect_to_email_login_url(email: str) -> HttpResponseRedirect:
     return HttpResponseRedirect(redirect_url)
 
 
+def realm_import_post_process(
+    request: HttpRequest,
+    confirmation_key: str,
+) -> HttpResponse:
+    try:
+        # We don't use get_object_from_key because we're fine if it's been marked used
+        confirmation = Confirmation.objects.get(
+            confirmation_key=confirmation_key, type__in=[Confirmation.REALM_CREATION]
+        )
+    except Confirmation.DoesNotExist:
+        raise JsonableError(_("Unauthenticated"))
+
+    preregistration_realm = confirmation.content_object
+    assert isinstance(preregistration_realm, PreregistrationRealm)
+    try:
+        realm = Realm.objects.get(string_id=preregistration_realm.string_id)
+    except Realm.DoesNotExist:
+        raise JsonableError(_("Requested organization doesn't exist."))
+
+    if not preregistration_realm.data_import_metadata["no_user_matching_email"]:
+        return HttpResponseRedirect(get_safe_redirect_to(reverse("login"), realm.url))
+
+    if request.method == "POST":
+        form = ImportRealmOwnerSelectionForm(request.POST)
+        if form.is_valid():
+            email = form.cleaned_data["email"]
+            user = UserProfile.objects.get(delivery_email__iexact=email, realm=realm)
+            user.delivery_email = preregistration_realm.email
+            user.role = UserProfile.ROLE_REALM_OWNER
+            user.save(update_fields=["delivery_email", "role"])
+            preregistration_realm.status = confirmation_settings.STATUS_USED
+            preregistration_realm.created_realm = realm
+            preregistration_realm.data_import_metadata["no_user_matching_email"] = False
+            preregistration_realm.save()
+            return HttpResponseRedirect(get_safe_redirect_to(reverse("login"), realm.url))
```

Comment:
Let's do this function inside `@transaction.atomic(durable=True)` on this function to avoid having to think about the possibility that this would be partially executed.

Reply:
Done.